### PR TITLE
Extension Chain optimization

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
I want to bring https://github.com/mono/monodevelop/pull/2491/commits/1706b02f558efacb778b63bf5f5c0b221e8d9a46 into discussion.

So, the idea of `FindNextImplementation` is to only return an item when there's at least one override/newslot definition or a serializable field/property. The question is _why_. It seems like wasted CPU for something that esentially saying: if the class is derived from the type.

Probably we should also check that the type is not the same, but it seems like pretty much pointless to do that check.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs [hunk 1]
```diff
@@ -387,15 +387,13 @@ void InitializeExtensionChain ()
...
-			foreach (var e in tempExtensions)
-				e.Dispose ();
+			extensionChain.Dispose ();
 
 			// Now create the final extension chain
 
 			extensions.Reverse ();
 			var defaultExts = CreateDefaultExtensions ().ToList ();
 			defaultExts.Reverse ();
-			extensionChain.SetDefaultInsertionPosition (defaultExts.FirstOrDefault ());
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs]: this was done for no reason, maybe it was supposed to be done on the new extension chain that was created?

**Comment by Therzok** [comment]
Adding DON'T MERGE YET, as I want to add unit tests for this.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs (hunk 1 above)
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs]: Yes, it should be called in the final extension chain. Not sure why this is not a problem right now.

Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs]: Pushed a fix.

**Comment by slluis** [comment]
The reflection check is done to avoid unnecessary chained calls. For example, with the current approach, if an extension overrides a single method of DotNetProjectExtension, only methods from DotNetProjectExtension will be chained through this instance. However, all methods from base classes (ProjectExtension, SolutionItemExtension, WorkspaceObjectExtension) will not include this instance in their chain. If we only check that the extension is a subclass, then methods from the classes at the top of the hierarchy will always be chained, since almost all extensions extensions will subclass them.

**Comment by Therzok** [comment]
Hmmm, I guess the best of both worlds then is to cache the methods that get reflected for each extension chain?

**Comment by slluis** [comment]
Yes, we can cache the info.

**Comment by Therzok** [comment]
Will revive by fixing it to maintain the old behaviour and caching the reflection checks.

**Pull request closed without merging**